make/compiler_flags

```diff
@@ -0,0 +1,190 @@
+# CXXFLAGS_EIGEN
+
+## GPU specific flags
+# CXXFLAGS_OPENCL
+# LDFLAGS_OPENCL
+# LDLIBS_OPENCL
+
+## MPI specific flags
+# CXXFLAGS_MPI
+# LDFLAGS_MPI
+# LDLIBS_MPI
+
+################################################################################
+# Update compiler flags with operating system specific modifications
+##
+ifeq ($(OS),Windows_NT)
+  CXXFLAGS_WARNINGS ?= -Wall -Wno-unused-function -Wno-uninitialized -Wno-unused-but-set-variable -Wno-unused-variable -Wno-sign-compare -Wno-unused-local-typedefs
+  CPPFLAGS_GTEST += -DGTEST_HAS_PTHREAD=0
+  CXXFLAGS_OS ?=
+  ## Detecting Process Bitness:
+  ## http://blogs.msdn.com/b/david.wang/archive/2006/03/26/howto-detect-process-bitness.aspx
+  ifeq ($(PROCESSOR_ARCHITECTURE)$(PROCESSOR_ARCHITEW6432),x86)
+    CXXFLAGS_OS += -m32
+  else
+    CXXFLAGS_OS += -m64
+  endif
+
+  LDLIBS ?= -static-libgcc
+  ifeq (gcc,$(CXX_TYPE))
+    LDLIBS += -static-libstdc++
+  endif
+
+  EXE := .exe
+endif
+
+ifeq ($(OS),Darwin)
+  ifeq (clang,$(CXX_TYPE))
+    CXXFLAGS_OS ?= -Wno-unknown-warning-option -Wno-tautological-compare -Wno-sign-compare
+  endif
+endif
+
+ifeq ($(OS),Linux)
+  CPPFLAGS_GTEST += -DGTEST_HAS_PTHREAD=0
+  CXXFLAGS_OS ?= -Wno-sign-compare
+  ifeq (gcc,$(CXX_TYPE))
+    CXXFLAGS_OS += -pthread
+  endif
+  ifeq (mingw32-gcc,$(CXX_TYPE))
+    LDLIBS ?= -static -lwinpthread
+    EXE := .exe
+    WINE ?= wine
+  endif
+  ifeq (clang,$(CXX_TYPE))
+    LDLIBS ?= -lpthread
+    CXXFLAGS += -stdlib=libc++
+  endif
+endif
+
+
+################################################################################
+# Setup GPU
+#
+# Adds the following to CXXFLAGS_OPENCL
+# OPENCL_DEVICE_ID - The ID of the GPU (default: 0)
+# OPENCL_PLATFORM_ID The ID of the OpenCL platform (default: 0)
+# Both IDs can be found through installing and calling clinfo
+
+ifdef STAN_OPENCL
+  ifeq ($(OS),Darwin)  # Mac OS X
+    LDLIBS_OPENCL ?= -framework OpenCL
+  else
+    LDLIBS_OPENCL ?= -lOpenCL
+  endif
+  OPENCL_DEVICE_ID ?= 0
+  OPENCL_PLATFORM_ID ?= 0
+
+  CPPFLAGS_OPENCL ?= -DSTAN_OPENCL -DOPENCL_DEVICE_ID=$(OPENCL_DEVICE_ID) -DOPENCL_PLATFORM_ID=$(OPENCL_PLATFORM_ID)
+  CXXFLAGS_OPENCL ?= -I $(OPENCL)
+endif
+
+################################################################################
+# Setup MPI
+#
+# Sets up MPI CXXFLAGS_MPI and LDLIBS_MPI to compile and link to MPI
+
+ifdef STAN_MPI
+  ifeq ($(OS),Windows_NT)
+    $(error MPI is not supported on Windows.)
+  endif
+
+  ifeq ($(OS),Darwin)
+    BOOST_LIBRARY_SUFFIX := .dylib
+  endif
+
+  ifeq ($(OS),Linux)
+    BOOST_LIBRARY_SUFFIX := .so
+  endif
+
+  MPI_TARGETS ?= $(addsuffix $(BOOST_LIBRARY_SUFFIX),$(BOOST)/stage/lib/libboost_serialization $(BOOST)/stage/lib/libboost_mpi) $(MPI_TEMPLATE_INSTANTIATION)
+  CPPFLAGS_MPI ?= -DSTAN_MPI
+
+  BOOST_LIBRARY_ABSOLUTE_PATH = $(abspath $(BOOST)/stage/lib)
+
+  LDLIBS_MPI ?= -Wl,-lboost_mpi -Wl,-lboost_serialization -Wl,-L,"$(BOOST_LIBRARY_ABSOLUTE_PATH)" -Wl,-rpath,"$(BOOST_LIBRARY_ABSOLUTE_PATH)"
```

Comment:
Out of my own experience on our cluster, I would recommend changing this to
`LDLIBS_MPI ?=-Wl,-L,"$(BOOST_LIBRARY_ABSOLUTE_PATH)" -Wl,-rpath,"$(BOOST_LIBRARY_ABSOLUTE_PATH)"`

In addition we need to make sure that the files listed `$MPI_TARGETS` are explicitly part of the linking command line (or add those to `LDLIBS_MPI`?). Not using the `-l` flags, but instead plugging in the file path's to the serialization and the mpi library ensures that no conflicting version of these libraries are picked up from the linker. At least doing these changes solved that problem for me on a system I work on which had boost serialization 1.53.0 in it's systems path's.